### ISSUE TITLE
avoidStruts isn't working on workspace 1

### ISSUE THREAD

**Issue by mrckndt** [xmonad/xmonad#15]
I'm using xmonad 0.12, xmonad-contrib 0.12  and xmobar on archlinux. After the upgrade from 0.11 to 0.12 avoidStruts isn't working on workspace 1. On all other workspaces my config works like expected.

my xmonad config

```
import XMonad

import System.Exit

import qualified XMonad.StackSet as W
import qualified Data.Map as M
import qualified XMonad.Actions.FlexibleResize as Flex

import XMonad.Util.Run
import XMonad.Actions.CycleWS
import Graphics.X11.ExtraTypes.XF86

import XMonad.Hooks.DynamicLog
import XMonad.Hooks.ManageDocks
import XMonad.Hooks.ManageHelpers
import XMonad.Hooks.EwmhDesktops
import XMonad.Hooks.SetWMName

import XMonad.Layout.Named
import XMonad.Layout.TrackFloating
import XMonad.Layout.ResizableTile


flexoKeyBindings conf@(XConfig {XMonad.modMask = modm}) = M.fromList $
    [ ((modm, xK_Return), spawn "termite"),
      ((modm, xK_r), spawn "rofi -terminal termite -modi run -show run"),
      ((modm .|. shiftMask, xK_Print), spawn "scrot"),
      ((modm, xK_Print), spawn "scrot -u"),

      ((0, xF86XK_LaunchA), spawn "xset dpms force off"),

      ((0, xF86XK_AudioRaiseVolume), spawn "ponymix --max-volume 100 increase 2"),
      ((0, xF86XK_AudioLowerVolume), spawn "ponymix --max-volume 100 decrease 2"),
      ((0, xF86XK_AudioMute), spawn "ponymix toggle"),
      ((0, xF86XK_MonBrightnessDown), spawn "xbacklight -dec 5"),
      ((0, xF86XK_MonBrightnessUp), spawn "xbacklight -inc 5"),

      ((modm .|. mod1Mask, xK_p), spawn "systemctl poweroff"),
      ((modm .|. mod1Mask, xK_r), spawn "systemctl reboot"),
      ((modm .|. mod1Mask, xK_l), spawn "slock"),

      ((modm .|. mod1Mask, xK_q), io (exitWith ExitSuccess)),

      ((modm, xK_q), kill),

      ((modm, xK_space ), sendMessage NextLayout),

      ((modm, xK_t), withFocused $ windows . W.sink),

      ((modm, xK_u), sendMessage (IncMasterN 1)),
      ((modm, xK_i), sendMessage (IncMasterN (-1))),


      -- focus
      ((modm, xK_n), windows W.focusDown),
      ((modm, xK_p), windows W.focusUp),
      ((modm, xK_Down), windows W.focusDown),
      ((modm, xK_Up), windows W.focusUp),

      ((modm, xK_comma), prevWS),
      ((modm, xK_period), nextWS),

      ((modm, xK_BackSpace), nextScreen),
      ((modm .|. shiftMask, xK_BackSpace), prevScreen),

      ((modm, xK_Escape), toggleWS),


      -- swap
      ((modm .|. shiftMask, xK_n), windows W.swapDown),
      ((modm .|. shiftMask, xK_p), windows W.swapUp),
      ((modm .|. shiftMask, xK_Down), windows W.swapDown),
      ((modm .|. shiftMask, xK_Up), windows W.swapUp),

      ((modm .|. shiftMask, xK_comma), shiftToPrev >> prevWS),
      ((modm .|. shiftMask, xK_period), shiftToNext >> nextWS),

      ((modm .|. shiftMask, xK_Return), windows W.swapMaster),


      -- resize
      ((modm .|. controlMask, xK_b), sendMessage Shrink),
      ((modm .|. controlMask, xK_f), sendMessage Expand),
      ((modm .|. controlMask, xK_n), sendMessage MirrorShrink),
      ((modm .|. controlMask, xK_p), sendMessage MirrorExpand),
      ((modm .|. controlMask, xK_Left), sendMessage Shrink),
      ((modm .|. controlMask, xK_Right), sendMessage Expand),
      ((modm .|. controlMask, xK_Down), sendMessage MirrorShrink),
      ((modm .|. controlMask, xK_Up), sendMessage MirrorExpand)
    ]

    ++

    [ ((m .|. modm, k), windows $ f i) | (i, k) <- zip (XMonad.workspaces conf) [xK_1,xK_2,xK_3,xK_4,xK_5,xK_6,xK_7,xK_8,xK_9,xK_0],
      (f, m) <- [(W.greedyView, 0), (W.shift, shiftMask)]
    ]


flexoMouseBindings (XConfig {XMonad.modMask = modm}) = M.fromList $
    [ ((modm, button1), (\w -> focus w >> mouseMoveWindow w >> windows W.shiftMaster)),
      ((modm, button3), (\w -> focus w >> Flex.mouseResizeWindow w))
    ]


flexoConfig = defaultConfig {
    normalBorderColor = "#aaaaaa",
    focusedBorderColor = "#ef2929",

    workspaces = ["1","2","3","4","5","6","7","8","9","10"],

    modMask = mod4Mask,
    keys = flexoKeyBindings,
    mouseBindings = flexoMouseBindings
}


main = do
    xmproc <- spawnPipe "xmobar ~/.xmonad/xmobar.hs"
    xmonad $ ewmh flexoConfig {
        logHook = do
            ewmhDesktopsLogHook >> setWMName "LG3D"
            dynamicLogWithPP xmobarPP {
                ppOutput = hPutStrLn xmproc,
                ppOrder = \(ws:l:t:[]) -> [l],
                ppLayout = xmobarColor "#ef2929" ""
            },

        startupHook = do
            ewmhDesktopsStartup
            spawn "hsetroot -fill $HOME/Dotfiles/background.png"
            spawn "xrdb $HOME/.Xresources",

        handleEventHook = do
            ewmhDesktopsEventHook
            docksEventHook
            fullscreenEventHook,

        manageHook =
            manageDocks <+>
            (isDialog --> doCenterFloat) <+>
            (isFullscreen --> doFullFloat) <+>
            composeAll [
                className =? "Pavucontrol" --> doCenterFloat,
                className =? "Eom" --> doCenterFloat,
                className =? "VirtualBox" --> doCenterFloat,
                className =? "Arandr" --> doCenterFloat,
                className =? "Remmina" --> doCenterFloat
            ],

        layoutHook = trackFloating $ avoidStruts $
            named "T" (ResizableTall 1 (2/100) (1/2) []) |||
            named "F" (Full)
}
```

my xmobar config

```
Config {
    font = "xft:Dejavu Sans Mono:size=10:antialias=true",
    bgColor = "#000000",
    fgColor = "#aaaaaa",
    position = Top,
    alignSep = "}{",
    commands = [
        Run Com "uname" ["-n"] "hostname" 0,
        Run Com "uname" ["-r"] "kernel" 0,

        Run BatteryN ["BAT0"] ["-t", "<left>"] 310 "BAT0",
        Run BatteryN ["BAT1"] ["-t", "<left>"] 310 "BAT1",

        Run Date "%a, %b %d %H:%M" "date" 100,
        Run StdinReader
    ],
    template = " <fc=#ef2929><icon=./.xmonad/heart.xbm/></fc> <fc=#ffffff>%hostname%</fc> %kernel%}{%StdinReader%  BAT <fc=#ffffff>%BAT0%</fc> <fc=#ffffff>%BAT1%</fc>  <fc=#ffffff>%date%</fc> "
}
```

Best,
Marco

**Comment by pjones** [comment]
Marco, when you upgraded did you start a fresh copy of xmonad or did you simply do a restart?

I've noticed that sometimes when I do a restart `autoStruts` stops working on the workspace I was on when I did the restart.  The solution is to use a key binding to disable then enable `autoStruts`. (I believe the problem is that xmobar is restarted and then placed on top of other windows before `autoStruts` kicks in.)

Try adding a key binding for `sendMessage ToggleStruts` to see if that helps.

**Comment by mrckndt** [comment]
I removed the object-file, binary and .hi and did a `xmonad --recompile` before starting the new xmonad-version.

**Comment by mrckndt** [comment]
After toggleing Struts it is working on all workspaces but that can't be the solution in my opinion. To be clear; the xmobar is hidden behind the windows and isn't on top of the other windows.

**Comment by pjones** [comment]
No, it's not a solution at all.  I just wanted to narrow down the problem.

It sounds like there's a timing issue when restarting xmonad and xmobar at the same time.  The solution might be to get `autoStruts` plugged into the startup hook.  But that's just a guess.

Patches welcome.

**Comment by jbiel** [comment]
I'm experiencing the same issue on the same platform as OP. The ToggleStruts workaround works for me.

**Comment by f1u77y** [comment]
Problem is in `fullscreenEventHook`. I don't know what is it, but anyway, for me your config without it hasn't this bug.

UPD: remove `fullscreenEventHook` and `isFulscreen --> doFullFloat` from your config and add fullscreenSupport from X.L.Fullscreen

**Comment by mrckndt** [comment]
Removing all the fullscreen stuff/hooks seems to "solve" the problem. But using the fullscreenEventHook from XMonad.Layout.Fullscreen isn't working either (no isFullscreen --> doFullFloat but XMonad.Layout.Fullscreen.fullscreenEventHook). So the problem seems to be another one.

**Comment by f1u77y** [comment]
`fullscreenEventHook` from X.L.Fullscreen is almost the same as in X.H.EwmhDesktops(maybe need to open another issue about cleanup). And use `fullscreenSupport`, not `fullscreenEventHook`. `fullscreenSupport` enables `fullscreenEventHook`, `fullscreenFull` layout modifier(you haven't got it in your config) and `fullscreenManageHook` that you had in your config. According to this, problem seems to be caused by lack of `fullscreenFull` in your config. To not forget to add something manually, use  `fullscreenSupport`.

**Comment by mrckndt** [comment]
Okay, with fullscreenSupport it is working for some reason. Because it is working and it seems to be a problem in contrib the issue can be closed.

**Comment by jakemco** [comment]
I have the same issue, but I use none of the `fullscreen*` options in my `xmonad.hs`. The `ToggleStruts` workaround works for me, but is there a better long term fix?

**Comment by mrckndt** [comment]
Maybe you should at least your post your config.

**Comment by jakemco** [comment]
``` haskell

import XMonad

import qualified XMonad.StackSet as W
import qualified Data.Map as M
import XMonad.Layout.IndependentScreens

import XMonad.Util.Run
import XMonad.Hooks.DynamicLog
import XMonad.Hooks.ManageDocks



main = do
    xmproc <- spawnPipe "xmobar"
    xmonad $ def
        { terminal    = myTerminal
        , modMask     = myModMask
        , workspaces  = myWorkspaces
        , keys        = myKeys
        , logHook = dynamicLogWithPP xmobarPP
            { ppOutput = hPutStrLn xmproc
            , ppTitle = xmobarColor "cyan" "" . shorten 50
            }
        , layoutHook = avoidStruts $ layoutHook def
        , manageHook = manageHook def <+> manageDocks
        }

myTerminal = "x-terminal-emulator"
myModMask = mod4Mask
myWorkspaces = withScreens 3 ["1","2","3","4","5","6","7","8","9"]

myKeys x = M.union (M.fromList (newKeys x)) (keys def x)

newKeys conf@(XConfig { XMonad.modMask = modm}) =
    [
     ((m .|. modm, k), windows $ onCurrentScreen f i)
         | (i,k) <- zip (workspaces' conf) [xK_1 .. xK_9]
         , (f,m) <- [(W.greedyView, 0), (W.shift, shiftMask)]
    ] ++ [
     ((m .|. modm, key), screenWorkspace sc >>= flip whenJust (windows . f))
         | (key, sc) <- zip [xK_w, xK_e, xK_r] [0..]
         , (f, m) <- [(W.view, 0), (W.shift, shiftMask)]
    ] ++ [
     ((mod4Mask .|. shiftMask, xK_l ), spawn "dm-tool lock")
    ] ++ [
     ((mod4Mask, xK_semicolon), sendMessage Expand)
    ] ++ [
     ((mod4Mask, xK_apostrophe), sendMessage ToggleStruts)
    ]

```

**Comment by f1u77y** [comment]
@jakemask see #18 but it's a bug
